feat(plugins): able to load plugin stages into UI

bpowell opened the pull request.

Commit:
feat(plugins): able to load plugin stages into UI

bpowell commented on app/index.deck:
```diff
@@ -186,6 +185,8 @@
     });
   </script>
 
+  <script crossorigin src="https://unpkg.com/react@16/umd/react.production.min.js"></script>
```
the reason for adding the React script tags is that when a react plugin is loaded, unless the plugin comes bundled with the react code, the plugin will fail to work because react is not on the page. i'm not really sure the best way to deal with this right now besides adding the script tags.

ncknt replied:
We need to find a different way (= adding it to package.json). It won't work in certain environment and unpkg.com is not a CDN and it's a security risk. I'm fairly certain this won't be merged.

bpowell replied:
👍 got that figured out!

bpowell commented on app/scripts/modules/core/src/registry/Registry.ts:
```diff
@@ -10,3 +10,5 @@ export class Registry {
   };
 }
+
+window.Registry = Registry;
```
not ideal either, but this will allow for plugins to use `Registry.pipeline.registerStage({...})` in their code and have the stage be loaded into the list of stages

ncknt replied:
This is indeed not ideal. Since plugins are loaded from within the app, just pass them the registry. You can require that a plugin has some type of init method that takes the registry as a parameter.

bpowell commented on plugin-loader.js:
```diff
@@ -0,0 +1,12 @@
+fetch('./settings-plugins.js')
```
fetches the plugins information that is laid down by halyard (as a k8s secret/configmap) and then appends the scripts to the page body

bpowell commented on settings-plugins.js:
```diff
@@ -0,0 +1,8 @@
+{
```
an example of what halyard might lay down for deck to consume to get the location of the plugins

bpowell commented on webpack.config.js:
```diff
@@ -27,6 +27,7 @@ function configure(env, webpackOpts) {
       settings: SETTINGS_PATH,
       'settings-local': './settings-local.js',
       app: './app/scripts/app.ts',
+      'plugin-loader': './plugin-loader.js',
```
adds the plugin loader to the page, AFTER the angularjs application is loaded. this way `Registry` is able to be used by the plugins

bpowell commented:
![image](https://user-images.githubusercontent.com/719460/63873992-414ee180-c986-11e9-88aa-3db73a078caf.png)

ncknt commented on app/index.deck:
```diff
@@ -12,8 +12,7 @@
 
-    .loading-placeholder {
-      position: absolute;
+    .loading-placeholder { position: absolute;
```
avoid that change since it's not needed.

bpowell replied:
👍

ncknt commented:
I'd tweak the approach and:
- Make the script available through Halyard (like you have it.)
- Stick a reference to the plugins in `settings-local.yml`. Could be the format you use in `settings-plugins.js` (still Halyard)
- Require that each script exposes a `function initPlugin(registry, (and more things...?)) {}`. entrypoint (e.g. CommonJS)
- In the app, add a `initializePlugins` function that loads scripts (e.g. with requireJS, almond etc.) and calls the `initPlugin(registry)`

Sorry can't expand too much now and I know you've been thinking about requirements so maybe it's off but it's also cleaner.

Commit:
fix(plugins): removed react scripts

Commit:
fix(plugins): no idea how this change actually happened

Commit:
stuff'n'things

Commit:
chore(plugins): load plugins w/ promise and script

Commit:
feat(plugins): Added plugin interfaces for developers to use

Commit:
fix(plugins): properly loads plugins from settings.js

Commit:
fix(plugins): removed uneeded code

Commit:
fix(plugins): change up interfaces

ncknt commented on app/scripts/app.ts:
```diff
@@ -15,6 +16,10 @@ import { ECS_MODULE } from '@spinnaker/ecs';
 import { AZURE_MODULE } from '@spinnaker/azure';
 
+initPlugins().catch(() => {
+  //TODO use CustomBanner to tell the user that plugin(s) have not been loaded
```
With this method, you could potentially, start the app before plugins are loaded. This might be ok for stages but that limits the ability of plugins to affect the lifecycle of the app.

To avoid it, `initPlugins().catch(notifytheuser).finally(starttheapp)`

ncknt commented:
Have you tried to make a plugin using the `@spinnaker/plugins` module? What's the workflow? Do I have to do provide options to exclude `@spinnaker/*` from my build?

As a side note, I find bizarre to have a module. I'd just expect a `.ts` for the interface. Sorry if it's not super accurate, I'll admit I just glanced at the config.

Commit:
fix(plugins): finally load app

bpowell commented:
Yes I have! It would look something like:
```
import * as React from 'react';
import { IPluginInitialize, IStageRegistry } from '@spinnaker/plugins';

class LikeButton extends React.Component {
  render() {
    return ( <div>Hello world, Hello Spinnaker</div> );
  }
}

function initialize(registry: IStageRegistry): void {
  registry.pipeline.registerStage({
    label: 'My Stage',
    description: 'My Stage',
    key: 'myStage',
    alias: 'myStage',
    addAliasToConfig: true,
    cloudProvider: 'kubernetes',
    component: LikeButton,
    supportsCustomTimeout: true,
    producesArtifacts: true,
  });
};

let init: IPluginInitialize = initialize;
const plugin = {
  initialize: init,
};

window.spinnakerSettings.onPluginLoaded(plugin);
```
As far as I know I don't think so. I can build the application without `@spinnaker/core` and it looks like it is not in the outputted javascript file.

If we just have a `.ts` file for the interface, then plugin developers would have to `import '@spinnaker/core'` which we don't want

ncknt commented on app/index.deck:
```diff
@@ -1,5 +1,5 @@
 <!doctype html>
-<html class="no-js" ng-app="netflix.spinnaker">
+<html class="no-js" >
```
Why the removal of `ng-app`?

ncknt replied:
Never mind, I scrolled down.

bpowell replied:
I did that based on this Stackoverflow post: https://stackoverflow.com/a/35250866

They said that Angularjs uses that to bootstrap it self and with the promise I was running into the issue where angularjs could not load the application.

ncknt commented:
Can you make a repo with a base plugin where all you do in import `@spinnaker/plugins` and can write a typed plugin? It would be helpful to understand the user experience and would also provide a template for plugin authors. I fear the `window.plugins.loadPlugin()` isn't that nice of an experience and being able to publish an ES module would be a nicer and more modern experience.

link108 commented on settings.js:
```diff
@@ -211,6 +211,7 @@ window.spinnakerSettings = {
     required: false,
   },
   pubsubProviders: ['google'], // TODO(joonlim): Add amazon once it is confirmed that amazon pub/sub works.
+  plugins: [],
```
can you update this to:
```suggestion
  plugins: {%plugins%},
```

ethanfrogers replied:
You'll probably need to add this to the `ISpinnakerSettings` interface for completeness sake.

link108 commented on app/scripts/modules/plugins/tsconfig.json:
```diff
@@ -0,0 +1,38 @@
+{
```
does this mean plugins must be typescript?

bpowell replied:
nope, i just used typescript to be like the rest of deck

link108 commented on app/scripts/modules/plugins/webpack.config.js:
```diff
@@ -0,0 +1,114 @@
+'use strict';
```
when does this script run? eg since plugins are loaded at runtime

bpowell replied:
runs at build time of deck

link108 replied:
what exactly is it doing / what is it for? My understanding is webpack is for compiling (maybe transpiling too), since plugins aren't available at build time, I'm not sure why we have this file. eg what is it compiling if nothing is there?

bpowell replied:
This is used to actually compile the module into a resulting library that can be consumed downstream by plugin developers

link108 replied:
hmmm I see you have some frontend portion of the stage-plugin, does that pull in this library for testing? or how does that work?

Commit:
feat(plugins): added halyard support

ethanfrogers commented:
@bpowell as someone without a ton of context on this implementation could you provide some deeper detail as to how the plugins will be loaded? What's the flow when the application starts up? Will plugin developers only need to import `@spinnaker/plugins` I think the answer is yes?. 

Again, I don't have the same context as those of you who have been working on this but I want to make sure we clearly communicate approach/intent when submitting upstream in absense of a formal design doc.

ethanfrogers commented:
I agree that calling `window.spinnakerSettings.onPluginLoaded` is a weird experience. Typically, calling anything on `window` isn't best practice because it's global state but, perhaps it's necessary in this case? What's the goal of `onPluginLoaded`?

ethanfrogers commented on app/scripts/modules/plugins/src/plugins.ts:
```diff
@@ -0,0 +1,13 @@
+
+declare global {
+  interface Window {
+    spinnakerSettings: any;
```
There's a `ISpinnakerSettings` interface defined in `app/scripts/modules/core/src/config/settings.ts`. Better to use that instead of `any`.

bpowell replied:
Wouldn't that mean that a plugin developer would have to import `@spinnaker/core` in order to use `window.spinnakerSettings.onPluginLoaded(plugin);`?

ethanfrogers replied:
I don't think so? Why would they? Is the goal for `@spinnaker/plugin` to not have a dependency on `@spinnaker/core`?

ethanfrogers replied:
Might be worth trying, at least.

bpowell replied:
Okay, I now see what you are saying. Coffee has kicked in!

bpowell replied:
`spinnakerSettings` is actually defined as `ISpinnakerSettings = (window as any).spinnakerSettings` so it has to be of type `any`. We get error TS2717 otherwise.

ethanfrogers commented on app/scripts/modules/plugins/tsconfig.json:
```diff
@@ -0,0 +1,38 @@
+    "typeRoots": ["../../../../node_modules/@types"],
+    "paths": {
+      "@spinnaker/core": ["../../core/lib"],
+      "core/*": ["core/src/*"]
```
I don't know much/anything about TS configs but `core/*` is defined as `../../core/lib/*` in the `kubernetes`  and `aws` modules. Not sure if defining it as `../../core/src/*` will cause any issues 🤷‍♂ .

bpowell replied:
I don't know. I copied it from the `docker` module

Commit:
fix(plugins): plugins is now an object

Commit:
fix(plugins): added plugin block to ISpinnakerSettings

bpowell commented:
Yes, a plugin developer should only have to import `@spinnaker/plugins`. The flow is before the application is actually loaded, if any plugins are defined in `settings.js`, they will be loaded by adding a script tag to the end of the page. And when that happens, the `Registry` is passed to an `initialize` method that the plugin developer creates that will register the stage. The plugin developer has to register the stage themselves. This gives the plugin developer more freedom in what they can do. Once all the plugins are loaded, the normal flow of how Spinnaker loads continues. If there are any errors in loading plugins, I'd like to use the custom banner stuff to show an error to the end user, but for right now, ignoring it or maybe `console.log` should be fine. No matter the status of loading plugins, the normal startup of Spinnaker will happen.

The goal of `onPluginLoaded` is to call the `initialize` method from the plugin to have it register itself as a stage.

link108 commented:
@bpowell tests are failing btw

link108 commented on halconfig/settings.js:
```diff
@@ -221,6 +221,7 @@ window.spinnakerSettings = {
     },
   },
   pubsubProviders: ['google'], // TODO(joonlim): Add amazon once it is confirmed that amazon pub/sub works.
+  plugins: ['%plugins%'],
```
This needs to match the other templates in the file
```suggestion
  plugins: {%plugins%},
```

link108 commented on app/scripts/plugin-loader.js:
```diff
@@ -0,0 +1,18 @@
+  });
+}
+
+export function initPlugins() {
```
can you leave a comment on how this works?

link108 replied:
talked with @bpowell about this, for our notes:

- want less about how this is working, eg don't need to know that it calls the function above it to do stuff, but more around how this is part of the larger project. 
- talk about how its the entry point for plugins

link108 commented:
@bpowell I'd like to see the description for the PR updated to what you are going to post to OSS.

Commit:
fix(tests): fixed broken tests

by removing ng-app, we placed the bootstrapping of angular when tests
were ran. bootstrapping normally occurs when the page is running in a
browser, not during tests. we moved the bootstrapping to a file that is
not being called during tests.

bpowell commented:
@link108 fixed tests!

Commit:
chore(plugins): added comments on plugin-loader

link108 committed:
chore(plugins): allow plugin templating

Commit:
chore(plugins): fixed up comments
